File: sdks/node/src/lazai/settlement.ts
import Web3 from "web3";

import {
  USER_HEADER,
  NONCE_HEADER,
  SIGNATURE_HEADER,
  FILE_ID_HEADER,
} from "./request";

export class SettlementSignature {
  constructor(
    public user: string,
    public nonce: BigInt,
    public signature: string,
    public file_id?: BigInt | null
  ) {}

  toRequestHeaders(): Record<string, string> {
    return {
      [USER_HEADER]: this.user,
      [NONCE_HEADER]: this.nonce.toString(),
      [SIGNATURE_HEADER]: this.signature,
      [FILE_ID_HEADER]:
        this.file_id !== undefined && this.file_id !== null
          ? this.file_id.toString()
          : "",
    };
  }
}

export class SettlementRequest {
  constructor(
    public nonce: BigInt,
    public user: string,
    public node: string,
    public file_id?: BigInt | null
  ) {}

  abiEncode() {
    const web3 = new Web3();
    return web3.eth.abi.encodeParameters(
      ["uint256", "address", "address"],
      [this.nonce, this.user, this.node]
    );
  }

  generateSignature(privateKey: string): SettlementSignature {
    const messageHash = Web3.utils.keccak256(this.abiEncode());
    const web3 = new Web3();
    const signedMessage = web3.eth.accounts.sign(messageHash, privateKey);
    return new SettlementSignature(
      this.user,
      this.nonce,
      signedMessage.signature,
      this.file_id
    );
  }
}

File: crates/lazai/src/settlement.rs
use crate::ClientError;
use crate::request::{FILE_ID_HEADER, NONCE_HEADER, SIGNATURE_HEADER, USER_HEADER};
use alith_data::wallet::LocalEthWallet;
use alloy::{
    primitives::{Address, U256, keccak256},
    sol,
    sol_types::SolType,
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

pub type SettlementTuple = sol! { tuple(uint256, address, address) };

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SettlementSignature {
    pub user: Address,
    pub nonce: u64,
    /// Hex-format signature
    pub signature: String,
    pub file_id: Option<U256>,
}

impl SettlementSignature {
    pub fn to_request_headers(&self) -> HashMap<String, String> {
        let mut headers = HashMap::new();
        headers.insert(USER_HEADER.to_string(), self.user.to_string());
        headers.insert(NONCE_HEADER.to_string(), self.nonce.to_string());
        headers.insert(SIGNATURE_HEADER.to_string(), self.signature.clone());

        if let Some(file_id) = self.file_id {
            headers.insert(FILE_ID_HEADER.to_string(), file_id.to_string());
        }

        headers
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct SettlementRequest {
    pub nonce: u64,
    pub user: Address,
    pub node: Address,
    pub file_id: Option<U256>,
}

impl SettlementRequest {
    #[inline]
    pub fn abi_encode(&self) -> Vec<u8> {
        SettlementTuple::abi_encode(&(U256::from(self.nonce), self.user, self.node))
    }

    pub async fn generate_signature(
        &self,
        wallet: &LocalEthWallet,
    ) -> Result<SettlementSignature, ClientError> {
        let encoded_data = self.abi_encode();
        let message_hash = keccak256(encoded_data);
        let signature = wallet
            .sign_message_hex(&message_hash.0)
            .await
            .map_err(|e| ClientError::SigningError(e.to_string()))?;

        Ok(SettlementSignature {
            user: self.user,
            nonce: self.nonce,
            signature: format!("0x{}", signature),
            file_id: self.file_id,
        })
    }
}
